token/confidential-transfer/elgamal-registry/src/instruction.rs

```diff
@@ -0,0 +1,196 @@
+        /// for the proof.
+        proof_instruction_offset: i8,
+    },
+}
+
+impl RegistryInstruction {
+    /// Unpacks a byte buffer into a `RegistryInstruction`
+    pub fn unpack(input: &[u8]) -> Result<Self, ProgramError> {
+        let (&tag, rest) = input
+            .split_first()
+            .ok_or(ProgramError::InvalidInstructionData)?;
+
+        Ok(match tag {
+            0 => {
+                let proof_instruction_offset =
+                    *rest.first().ok_or(ProgramError::InvalidInstructionData)?;
+                Self::CreateRegistry {
+                    proof_instruction_offset: proof_instruction_offset as i8,
+                }
+            }
+            1 => {
+                let proof_instruction_offset =
+                    *rest.first().ok_or(ProgramError::InvalidInstructionData)?;
+                Self::UpdateRegistry {
+                    proof_instruction_offset: proof_instruction_offset as i8,
+                }
+            }
+            _ => return Err(ProgramError::InvalidInstructionData),
+        })
+    }
+
+    /// Packs a `RegistryInstruction` into a byte buffer.
+    pub fn pack(&self) -> Vec<u8> {
+        let mut buf = vec![];
+        match self {
+            Self::CreateRegistry {
+                proof_instruction_offset,
+            } => {
+                buf.push(0);
+                buf.extend_from_slice(&proof_instruction_offset.to_le_bytes());
+            }
+            Self::UpdateRegistry {
+                proof_instruction_offset,
+            } => {
+                buf.push(1);
+                buf.extend_from_slice(&proof_instruction_offset.to_le_bytes());
+            }
+        };
+        buf
+    }
+}
+
+/// Create a `RegistryInstruction::CreateRegistry` instruction
+pub fn create_registry(
+    funding_address: &Pubkey,
+    owner_address: &Pubkey,
+    proof_location: ProofLocation<PubkeyValidityProofData>,
+) -> Result<Vec<Instruction>, ProgramError> {
+    let elgamal_registry_address = get_elgamal_registry_address(owner_address, &id());
+
+    let mut accounts = vec![
+        AccountMeta::new(*funding_address, true),
+        AccountMeta::new(elgamal_registry_address, false),
+        AccountMeta::new_readonly(*owner_address, false),
+        AccountMeta::new_readonly(system_program::id(), false),
+    ];
+    let proof_instruction_offset = proof_instruction_offset(&mut accounts, proof_location);
+
+    let registry_instruction = Instruction {
+        program_id: id(),
+        accounts,
+        data: RegistryInstruction::CreateRegistry {
+            proof_instruction_offset,
+        }
+        .pack(),
+    };
+    append_zk_elgamal_proof(registry_instruction, proof_location)
+}
+
+/// Create a `RegistryInstruction::UpdateRegistry` instruction
+pub fn update_registry(
+    registry_account: &Pubkey,
+    owner: &Pubkey,
+    proof_location: ProofLocation<PubkeyValidityProofData>,
+) -> Result<Vec<Instruction>, ProgramError> {
```

Comment:
nit: It's up to you, but there's an inconsistency between the two instruction creators. The first derives the address for you, but this one doesn't. I would say it's probably easier to have them all derive the address for you, so this doesn't need to take the `registry_account`

Reply:
Yep, good idea!